### DISCUSSION TITLE
Consider making tesseract_planning its own repo.

### DISCUSSION THREAD

**Discussion by mpowelson** [tesseract-robotics/tesseract#421] [Ideas]
We already split tesseract_ros/ros2 into their own repos. tesseract_planning has a lot of dependencies that core tesseract does not need. Further, I would expect that tesseract_planning will continue to be under significant development for a while, but hopefully core tesseract will be more stable. I see this as particularly useful if we can address #419. Then developers could build just tesseract_planning.

**Comment by mpowelson** [comment]
More than that, we should consider making each directory in tesseract_motion_planners its own package. The current structure takes a very long time to build when building especially with Clang-tidy enabled. The additional parallelization would probably improve build times overall  since they all have very different dependencies.

**Comment by Levi-Armstrong** [comment]
@johnwason, We have been discussing breaking up this repository into three repositories. The idea is to to move tesseract_planning into its own repository and move tesseract_python into its own repository. Do you have any issues with this? The idea is that the remaining packages in tesseract repository are more stable than the others.

**Comment by johnwason** [comment]
@Levi-Armstrong I don't see any issue with this. tesseract_python is already in its own directory.

**Comment by Levi-Armstrong** [comment]
@mpowelson, @johnwason, @marip8 Tomorrow I plan on breaking out tesseract_planning and tesseract_python into its own repository. Should we also breakout tesseract_visualization into its own repository? I can see this growing into multiple packages possibly in the future.

**Reply by marip8** [reply]
I think that depends on what you have planned for `tesseract_visualization`. I think the current content (less the `ignition` components) should stay in the main `tesseract` repository because it defines interfaces and provides a trajectory viewer that is generic to most users. If your plan is to add a lot of implementations of the visualization interfaces with `ignition` or something else, then I think the implementation-specific code would be appropriate to put in a separate repository. 

Sidebar: it appears that `tesseract_visualization` depends on the `Instruction` class from `tesseract_planning`. I think it would also be valuable to define the core `tesseract_command_language` features (i.e. `Instruction`, `CompositeInstruction`, possibly the XML parsing aspects, etc.) in the main `tesseract` repository, rather than in `tesseract_planning`, especially if other modules depend on those interfaces. Keeping that in the main repository would also allow you and others to use the command language capability without having to download and compile all of the planning code

**Reply by mpowelson** [reply]
Yeah I think I'd leave tesseract_visualization in the main repo for now based on what is currently in it. Implementations could go in their own repos.  Moving all of tesseract_command_language into the main repo might make sense, but I do see that as more experimental than the rest of tesseract core. Presumably you could also use type erasure in tesseract_visualization if you wanted to eliminate that dependency.

**Reply by Levi-Armstrong** [reply]
What do you all think about moving the `Instruction.h` and `Waypoint.h` to tesseract_common but namespace it under `tesseract_planning`?

**Reply by Levi-Armstrong** [reply]
Or it could go under tesseract_common namespace and I can retype it under the tesseract_planning namespace.

**Reply by johnwason** [reply]
Mixing namespaces would not make much sense for the Python wrappers, since they are organized by module. Regardless of what their C++ namespace is, they would be in the tesseract_common module. Would it make more sense to put them in tesseract_kinematics?

**Reply by Levi-Armstrong** [reply]
It looks like the easiest will be to move tesseract_command_language to the stay in the core tesseract repository. I have created a PR against tesseract and once merged I will force push to tesseract_planning and tesseract_python using the approach below to possibly reduce commits. Let me know if you see an y issues.

**Reply by johnwason** [reply]
So move tesseract_command_language but keep it as its own package? That should be fine.

I'd like to keep my commit history for tesseract_python.

**Reply by Levi-Armstrong** [reply]
> So move tesseract_command_language but keep it as its own package? That should be fine.

That is what will be done.

> I'd like to keep my commit history for tesseract_python.

Yea, it should not delete any commits. It looks like it will just keep commits related to the folder being moved. So some will remain in tesseract and others will remain in tesseract_python.

**Reply by marip8** [reply]
What exactly is the scope of the command language? Is the purpose to be a general, type-erased, "nest-able", easily override-able data representation that can be parsed into and out of XML? If so and several of the other modules will utilize it (i.e. visualization, etc.), then I would think that it belongs in the main repository rather than its own repository. Then the extensions of the command language (like for motion/process planning) can go in their own respective repositories

**Comment by Levi-Armstrong** [comment]
Everyone, the following new repositories have been created and the code has been pushed there. I will create PR against Tesseract removing them. If it is decided to break out tesseract_visualization I can make that change then.

[Tesseract Planning](https://github.com/ros-industrial-consortium/tesseract_planning)
[Tesseract Python](https://github.com/ros-industrial-consortium/tesseract_python)

**Reply by mpowelson** [reply]
I know for tesseract_ros we just deleted the old code, however I was just looking into other ways of doing this. Do you think we should use [this method](https://docs.github.com/en/free-pro-team@latest/github/using-git/splitting-a-subfolder-out-into-a-new-repository)? It looks like it gets rid of commits that don't include code in the new repo.

**Reply by Levi-Armstrong** [reply]
I am fine with that. Do you want to give it a shot and push to a different branch?

**Reply by mpowelson** [reply]
[Apparently](https://stackoverflow.com/questions/61127751/difference-between-git-filter-branch-subdirectory-filter-git-subtree-split-p) the method I linked before is using deprecated tools but should give the same results.

I used this instead. Taken from [here](https://stackoverflow.com/questions/359424/detach-move-subdirectory-into-separate-git-repository/17864475#17864475)

```
git subtree split -P <name-of-folder> -b <name-of-new-branch>
```

I pushed it [here](https://github.com/ros-industrial-consortium/tesseract_planning/tree/tesseract_planning_repo)

**Reply by mpowelson** [reply]
It did get rid of a bunch of commits. However it looks like a lot of the early ones (including bullet, tesseract_ros, etc) are still around.

**Reply by Levi-Armstrong** [reply]
@mpowelson and @johnwason The tesseract_python package has been broken out into its own repository and the git subtree split worked well.

**Reply by gavanderhoorn** [reply]
you're probably aware, but just making sure: `subtree` splits result in rewritten history. Any references to commits will no longer work.

**Reply by Levi-Armstrong** [reply]
@gavanderhoorn Thanks, I was not sure what it did. I did not run the split on the tesseract repo but just deleted the the directories to avoid any issues.

**Comment by Levi-Armstrong** [comment]
@mpowelson and @johnwason,

Everything has been broken out into its own repository and CI along with rosinstall files have been updated. Also tesseract_ext is no longer needed for noetic on windows. Though tesseract_python still required it because swig 4.0.2 has not been release yet.

Also @johnwason, I gave you write access to tesseract_python.
